docs/designs/custom_allreduce_training_loop_support.md

```diff
@@ -0,0 +1,404 @@
+# Support Customizing the Training Loop Using AllReduce
+
+This document describes the design for supporting customizing
+the training loop when users use the AllReduce distribution strategy.
+
+## Motivation
+
+Users need to define the forward computation, loss function,
+optimizer, and dataset function in ElasticDL. ElasticDL provides
+the training loop with those definitions. ElasticDL only supports
+Keras API to define the forward computation. It maybe not flexible for
+users to define complex models in CV or NLP. Sometimes, users need to
+customize the training loop to control the model iteration. For example,
+For example, given a pre-trained network, users might only want to
+optimize a new set of output weights.
+
+If ElasticDL supports customizing the training loop, users can use different
+deep-learning library (e.g., TensorFlow, Pytorch) to define their training
+loop. ElasticDL only need to provide dynamic data partitioning for
+dataset and elastic AllReduce to merge gradients across workers.
+
+## The Training Loop of ElasticDL
+
+The training loop of ElasticDL on the worker contains two steps:
+
+- Get tasks from the master to create dataset with data shards in tasks.
+- Execute training loop to calculate gradients and update the model
+using the dataset.
+
+![Training Loop](../images/training_loop.jpg)
+
+From the figure, elastic training of ElasticDL focus on data partitioning
```

Comment:
This part is not very clear. We can rewrite it:

In order to support customized training loop with ElasticDL AllReduce, ElasticDL provides users with a dataset with dynamic partition support, and an API `elastic_run` for elastic training. 

Users need to implement their own training logic for a training data batch in a function `train_one_batch`:

```Python
def train_one_batch(batch_data):
    loss = forward(batch_data)
    gradients = backward(loss)
    gradients = allreduce(gradients)
    update_model(gradients)
```

Then users can write the training loop using `train_one_batch`, the dataset and `elastic_run` API  provided by ElasticDL.

```python
def train(dataset, elastic_run):
    """
    Arguments:
        dataset: tf.data.Dataset. ElasticDL creates a dataset according to
            tasks from the master
        elastic_manager: A manager to execute elastic training.
    """
    for batch_data in dataset:
        elastic_train = elastic_run(train_one_batch)
        elastic_train(batch_data)
```